File: canrisk-bridge/python/tests/test_canrisk_client.py
from pathlib import Path

import pytest

from canrisk_bridge.canrisk_client import (
    CANRISK_HEADER,
    ApiError,
    CanRiskClient,
    CanRiskFormatError,
)

_PEDIGREE_ID = "11111111-1111-1111-1111-111111111111"

FIXTURES = Path(__file__).resolve().parents[2] / "fixtures"


def _sample_canrisk() -> str:
    return (FIXTURES / "sample-canrisk.txt").read_text(encoding="utf-8")


class _StubResponse:
    def __init__(self, status_code: int, text: str) -> None:
        self.status_code = status_code
        self.text = text


class _RecordingGateway:
    def __init__(self, response: _StubResponse) -> None:
        self._response = response
        self.last_url: str = ""
        self.last_headers: dict[str, str] = {}

    def get_text(self, url: str, *, headers: dict[str, str]) -> _StubResponse:
        self.last_url = url
        self.last_headers = headers
        return self._response


def _client(gateway: _RecordingGateway) -> CanRiskClient:
    return CanRiskClient(base_url="https://evagene.example", api_key="evg_test", http=gateway)


def test_gets_canrisk_endpoint_with_correct_headers() -> None:
    gateway = _RecordingGateway(_StubResponse(200, _sample_canrisk()))

    body = _client(gateway).fetch(_PEDIGREE_ID)

    assert gateway.last_url == (
        f"https://evagene.example/api/pedigrees/{_PEDIGREE_ID}/risk/canrisk"
    )
    assert gateway.last_headers["X-API-Key"] == "evg_test"
    assert gateway.last_headers["Accept"] == "text/tab-separated-values"
    assert body.startswith(CANRISK_HEADER)


def test_raises_api_error_on_non_2xx() -> None:
    gateway = _RecordingGateway(_StubResponse(500, ""))

    with pytest.raises(ApiError):
        _client(gateway).fetch(_PEDIGREE_ID)


def test_raises_format_error_when_header_missing() -> None:
    gateway = _RecordingGateway(_StubResponse(200, "not a canrisk file"))

    with pytest.raises(CanRiskFormatError):
        _client(gateway).fetch(_PEDIGREE_ID)

File: canrisk-bridge/r/tests/testthat/test-canrisk_client.R
PEDIGREE_ID <- "11111111-1111-1111-1111-111111111111"

sample_canrisk <- function() {
  path <- normalizePath(
    file.path("..", "..", "..", "fixtures", "sample-canrisk.txt"),
    mustWork = TRUE
  )
  paste(readLines(path, warn = FALSE), collapse = "\n")
}

recording_gateway <- function(status, body) {
  last <- new.env(parent = emptyenv())
  last$url <- NULL
  last$headers <- NULL

  gateway <- function(request) {
    last$url <- request$url
    last$headers <- request$headers
    list(status = status, body = body)
  }
  list(gateway = gateway, last = last)
}

test_that("GETs the canrisk endpoint with the documented headers", {
  recorder <- recording_gateway(status = 200L, body = sample_canrisk())

  body <- fetch_canrisk(
    recorder$gateway,
    base_url = "https://evagene.example",
    api_key = "evg_test",
    pedigree_id = PEDIGREE_ID
  )

  expect_equal(
    recorder$last$url,
    sprintf("https://evagene.example/api/pedigrees/%s/risk/canrisk", PEDIGREE_ID)
  )
  expect_equal(unname(recorder$last$headers[["X-API-Key"]]), "evg_test")
  expect_equal(unname(recorder$last$headers[["Accept"]]), "text/tab-separated-values")
  expect_true(startsWith(body, CANRISK_HEADER))
})

test_that("raises api_error on non-2xx status", {
  recorder <- recording_gateway(status = 500L, body = "")

  expect_error(
    fetch_canrisk(recorder$gateway, "https://evagene.example", "evg_test", PEDIGREE_ID),
    class = "api_error"
  )
})

test_that("raises canrisk_format_error when header is missing", {
  recorder <- recording_gateway(status = 200L, body = "not a canrisk file")

  expect_error(
    fetch_canrisk(recorder$gateway, "https://evagene.example", "evg_test", PEDIGREE_ID),
    class = "canrisk_format_error"
  )
})
